src/linear_algebra/cross_product.lean

```diff
@@ -0,0 +1,116 @@
+
+* `×₃` for the cross product
+
+## Tags
+
+crossproduct
+-/
+
+open_locale vectors
+
+variables {R : Type*} [comm_ring R]
+
+/-- The cross product of two vectors in $R^3$ for $R$ a commutative ring. -/
+def cross_product : (fin 3 → R) →ₗ[R] (fin 3 → R) →ₗ[R] (fin 3 → R) :=
+begin
+  apply linear_map.mk₂ R (λ (a b : fin 3 → R),
+    ![ (a 1)*(b 2) - (a 2)*(b 1) ,
+       (a 2)*(b 0) - (a 0)*(b 2) ,
+       (a 0)*(b 1) - (a 1)*(b 0) ]);
+  intros;
+  simp only [vec3_add,
+    pi.add_apply, smul_eq_mul, matrix.smul_cons, matrix.smul_empty, pi.smul_apply];
+  apply vec3_eq;
+  ring,
+end
+
+localized "infixl ` ×₃ `: 68 := cross_product" in vectors
+
+lemma cross_product_apply (a b : fin 3 → R) :
+  a ×₃ b =
+    ![ (a 1)*(b 2) - (a 2)*(b 1) ,
+       (a 2)*(b 0) - (a 0)*(b 2) ,
+       (a 0)*(b 1) - (a 1)*(b 0) ] :=
+rfl
+
+lemma cross_product_anticomm (v w : fin 3 → R) :
+  - (v ×₃ w) = w ×₃ v :=
+by simp [cross_product_apply, mul_comm]
+
+lemma cross_product_anticomm' (v w : fin 3 → R) :
+  v ×₃ w + w ×₃ v = 0 :=
+by rw [add_eq_zero_iff_eq_neg, cross_product_anticomm]
+
+lemma cross_product_self (v : fin 3 → R) :
+  v ×₃ v = 0 :=
+by simp [cross_product_apply, mul_comm]
+
+/-- The cross product of two vectors is perpendicular to the first vector. -/
+lemma dot_self_cross_product (v w : fin 3 → R) :
+  v ⬝ (v ×₃ w) = 0 :=
+by simp [cross_product_apply, vec3_dot_product, mul_sub, mul_assoc, mul_left_comm]
+
+/-- The cross product of two vectors is perpendicular to the second vector. -/
+lemma dot_cross_product_self (v w : fin 3 → R) :
+  w ⬝ (v ×₃ w) = 0 :=
+by rw [← cross_product_anticomm, matrix.dot_product_neg, dot_self_cross_product, neg_zero]
+
+/-- Cyclic permutations preserve the triple product. See also `triple_product_eq_det`. -/
+lemma triple_product_permutation (u v w : fin 3 → R) :
+  u ⬝ (v ×₃ w) = v ⬝ (w ×₃ u) :=
+begin
+  simp only [cross_product_apply, vec3_dot_product,
+    matrix.head_cons, matrix.cons_vec_bit0_eq_alt0, matrix.empty_append, matrix.cons_val_one,
+    matrix.cons_vec_alt0, matrix.cons_append, matrix.cons_val_zero],
+  ring,
+end
+
+/-- The triple product of `u`, `v`, and `w` is equal to the determinant of the matrix
+    with those vectors as its rows. -/
+theorem triple_product_eq_det (u v w : fin 3 → R) :
+  u ⬝ (v ×₃ w) = matrix.det ![u, v, w] :=
+begin
+  simp only [vec3_dot_product, cross_product_apply, matrix.det_fin_three,
+    matrix.head_cons, matrix.cons_vec_bit0_eq_alt0, matrix.empty_vec_alt0, matrix.cons_vec_alt0,
+    matrix.vec_head_vec_alt0, fin.fin_append_apply_zero, matrix.empty_append, matrix.cons_append,
+    matrix.cons_val', matrix.cons_val_one, matrix.cons_val_zero],
+  ring,
+end
+
+/-- For a cross product of three vectors, their sum over the three even permutations is equal
+    to the zero vector. -/
+theorem jacobi_identity (u v w : fin 3 → R) :
+  u ×₃ (v ×₃ w) + v ×₃ (w ×₃ u) + w ×₃ (u ×₃ v) = 0 :=
+begin
+  repeat {rw cross_product_apply},
+  norm_num,
+  ring_nf,
+  tauto,
+end
```

Comment:
Here's another way to prove this:

```suggestion
instance : lie_ring (fin 3 → R) :=
{ bracket := λ u v, u ×₃ v,
  add_lie := linear_map.map_add₂ _,
  lie_add := λ u, linear_map.map_add _,
  lie_self := cross_product_self,
  leibniz_lie := λ u v w, begin
    dsimp only [has_bracket.bracket, cross_product_apply],
    ext i,
    fin_cases i; norm_num; ring,
  end,
  ..pi.add_comm_group }

/-- For a cross product of three vectors, their sum over the three even permutations is equal
    to the zero vector. -/
theorem jacobi_identity (u v w : fin 3 → R) :
  u ×₃ (v ×₃ w) + v ×₃ (w ×₃ u) + w ×₃ (u ×₃ v) = 0 :=
lie_jacobi u v w
```

Reply:
Thank you for your suggestion!

What is the `liebniz_lie` thingie? I looked into `src/algebra/lie/basic.lean` and it says the following.
```
(leibniz_lie : ∀ (x y z : L), ⁅x, ⁅y, z⁆⁆ = ⁅⁅x, y⁆, z⁆ + ⁅y, ⁅x, z⁆⁆)
```
Does mathlib use a different definition of Lie ring than the definition from wikipedia?

Reply:
[Zulip isn't sure whether my code above is a good idea](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/cross.20product/near/266799595).

> Does mathlib use a different definition of Lie ring than the definition from wikipedia?

I would guess that @ocfnash (or his paper at https://arxiv.org/abs/2112.04570) can answer that.

Reply:
Nope, we use the same definition though as Eric rightly observes you probably want to look at section 2 of that article.

Reply:
The paper says: "In the presence of the other Lie algebra axioms, each of the following are equivalent:
(2.1) jacobi identity (2.2) liebniz lie (2.3) something"

Are they be equivalent in the presence of other Lie ring axioms only?

Reply:
Yes.

Reply:
Thanks!

Reply:
The `leibniz_lie` axiom is the "right" formulation. An important operation in Lie algebra is the `[x, -]` adjoint action, and what it's saying is that the Lie bracket is a homomorphism of Lie algebra representations (from `A \otimes A ` to `A`).  Writing `ad x` for `y \mapsto [x, -]`, it's
```
ad x [y, z] = [ad x y, z] + [y, ad x z]
```

Reply:
Based on the Zulip discussion started by @eric-wieser, maybe it would be better to have this `instance` be a `def` and then use it as a local instance?

It turns out there's another Lie algebra instance on pi types like this for an abelian Lie algebra structure (i.e., the zero bracket) and it would be good to not conflict with it.

Reply:
Oops, I should have looked at the paper -- my comment about `ad` is already there.

Reply:
Shall we keep the Jacobi identity without the connection to Lie rings in the PR then?

Reply:
The Lie ring structure is still a good source of free lemmas, so we should keep it. Here's (untested) code for how I mean to restructure @eric-wieser's code:
```lean
def cross_product.lie_ring : lie_ring (fin 3 → R) :=
{ bracket := λ u v, u ×₃ v,
  add_lie := linear_map.map_add₂ _,
  lie_add := λ u, linear_map.map_add _,
  lie_self := cross_product_self,
  leibniz_lie := λ u v w, begin
    dsimp only [has_bracket.bracket, cross_product_apply],
    ext i,
    fin_cases i; norm_num; ring,
  end,
  ..pi.add_comm_group }

local attribute [instance] cross_product.lie_ring

/-- For a cross product of three vectors, their sum over the three even permutations is equal
    to the zero vector. -/
theorem jacobi_identity (u v w : fin 3 → R) :
  u ×₃ (v ×₃ w) + v ×₃ (w ×₃ u) + w ×₃ (u ×₃ v) = 0 :=
lie_jacobi u v w
```

Reply:
I think it might make sense to give `leibniz_lie` its own lemma, perhaps `leibniz_cross` or something, rather than proving it in  `cross_product.lie_ring`.

Reply:
Good! Can I do the refactoring as you both suggested?